NOISSUE - Docs for domains , groups, channel, things relationship (#174)
absmach/magistrala-docs: docs/authorization.md

# Authorization

Magistrala allows for fine-grained control over user permissions, taking into account hierarchical relationships between entities domains, groups, channels, and things. The structure and functionality of an authorization system implemented using [SpiceDB](https://github.com/authzed/spicedb) and its associated [schema language](https://authzed.com/docs/reference/schema-lang). `auth` service backed by SpiceDB manages permissions for users, domains, groups, channels, and things.

## Domains

> Domain contains **Things**, **Channels**, and **Groups**. **Users** can be a member of a Domain with different types of available relations, These relation provides access control to the entities in the domain.

Reviewer: ```suggestion
Domain contains **Things**, **Channels**, and **Groups**. A **User** can be a member of a Domain with different types of available relations, These relation provides access control to the entities in the domain.
```

> Domain contains **Things**, **Channels**, and **Groups**. A **User** can be a member of a Domain with different types of available relations, These relation provides access control to the entities in the domain.

Reviewer: ```suggestion
Domain contains **Things**, **Channels**, and **Groups**. A **User** can be a member of a Domain with different types of available relations. This relation provides access control to the entities in the domain.
```

style Domain stroke-width:3px,margin-top:10px,margin-bottom:10px
```

#### Domain Entities Relations

Domain holds entities such as Groups, Channels, and Things.
The entities created in a domain don't have any hierarchical structure between them.

> Example: In domain_1 a user creates the following entities `group_1`, `group_2`, `thing_1`, `thing_2`, `channel_1`, `channel_2`. By default, there is no relation between the entities, until the user assigns a relation between the entities

Reviewer: ```suggestion
Example: In `domain_1` a user creates the following entities `group_1`, `group_2`, `thing_1`, `thing_2`, `channel_1`, `channel_2`. By default, there is no relation between the entities, until the user assigns a relation between the entities
```

```mermaid
graph
   subgraph domain_1
      direction TB
      Gr1["group_1"]
      Gr2["group_2"]

      Th1["thing_1"]
      Th2["thing_2"]

      Ch1["channel_1"]
      Ch2["channel_2"]

   end
```

##### Channel Thing Connect/Disconnect

> `Thing` represents a device (or an application) connected to Magistrala that uses the platform for message exchange with other `things`.

Reviewer: Also update the documentation here  https://docs.magistrala.abstractmachines.fr/entities/#things-service

`Channel` is a message conduit between things connected to it. It serves as a message topic that can be consumed by all of the things connected to it.
Things can publish or subscribe to the channel.

Thing and channel can be connected to multiple channels using the following API.

```bash
curl -sSiX POST http://localhost/connect -H "Content-Type: application/json" -H "Authorization: Bearer <domain_user_access_token>" -d @- << EOF
{  
  "thing_id": "<thing_id>",
  "channel_id": "<channel_id>"
}  
EOF 
```

_*The below diagram shows `thing_1` is connected to `channel_1` and `channel_2` , then `thing_2` is connected to `channel_2`. This relationship can be established using the provided request*_

```mermaid
graph
   subgraph domain_1
      direction BT
      Gr1["group_1"]
      Gr2["group_2"]

      Th1["thing_1"]
      Th2["thing_2"]

      Ch1["channel_1"]
      Ch2["channel_2"]

      Th1 --->|connect| Ch1
      Th1 --->|connect| Ch2

      Th2 --->|connect| Ch2
   end
```

##### Channel Group Relation

> A group serves as a parent entity that can contain both groups and channels as children. Child groups, in turn, can consist of further child groups or channels, forming a nested hierarchy. Notably, channels, which are distinct entities, cannot have child channels but can connect to multiple things. The concept of parentage signifies the relationship between higher-level entities and their subordinate components. Ancestors in this system refer to entities higher up in the hierarchy, and while a child group can have multiple ancestors, a channel can only belong to a single parent group. This hierarchical arrangement provides a structured and organized framework for managing information within the magistrala.

Reviewer 2: ...within the Magistrala.

Reviewer 2: Always capitalize `Magistrala` in other places, as well.

style Domain stroke-width:3px,margin-top:10px,margin-bottom:10px
```

#### Domain Entities Relations

Domain holds entities such as Groups, Channels, and Things.

> The entities created in Domain don't have any hierarchical structure between them.

Reviewer: ```suggestion
The entities created in a domain don't have any hierarchical structure between them.
```

> Example: In Domain_1 a user creates the following entities Group 1, Group_2, Thing 1, Thing 2, Channel 1, Channel 2. By default, there is no relation between the entities, until the user assigns a relation between the entities

Reviewer: Let's be consistent with naming i.e either use snake case every where.
```suggestion
Example: In domain_1 a user creates the following entities group_1, group_2, thing_1, thing_2, channel_1, channel_2. By default, there is no relation between the entities, until the user assigns a relation between the entities
```

Reword the second line because when creating an entity we add a policy hence there is a relation between the entity and the user.

Author: Yes, While creating an entity there will be relation between entity and user. 
But here the case different, We want to explain to reader about relation between a entity with other entities. 
Since this part of document explain about entity relationship, and didn't mentioned about user entity relation. 
User entity relation in explained in detail in following parts of documents.

```mermaid
graph
   subgraph Domain_1
      direction TB
      Gr1["Group 1"]
      Gr2["Group 2"]

      Th1["Thing 1"]
      Th2["Thing 2"]

      Ch1["Channel 1"]
      Ch2["Channel 2"]

   end
```

##### Channel Thing Connect/Disconnect

> `Thing` represents a device (or applications) connected to Magistrala that uses the platform for message exchange with other `things`.

Reviewer: ```suggestion
`Thing` represents a device (or an application) connected to Magistrala that uses the platform for message exchange with other `things`.
```

`Channel` is a message conduit between things connected to it. It serves as a message topic that can be consumed by all of the things connected to it.
Things can publish or subscribe to the Channel.

Thing and Channel can be connected to multiple channels using the following API.

```bash

> curl --location 'http://localhost/connect' \

Reviewer: Use same structure as used in the API docs

```sh
curl -sSiX POST http://localhost/connect -H "Content-Type: application/json" -H "Authorization: Bearer <domain_user_access_token>" -d @- << EOF
{
  "thing_id": "<Thing 1 ID>",
  "channel_id": "<Channel 1 ID>"
}
EOF
```

--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'Authorization: Bearer <DOMAIN_USER_ACCESS_TOKEN>' \
--data '{
  "thing_id": "<Thing 1 ID>",
  "channel_id": "<Channel 1 ID>"
}'
```

> _*The below diagram shows Thing 1 connect to channel 1 , channel 2 and Thing 2 connect to Thing 2. This relationship can be established using the provided request*_

Reviewer: ```suggestion
_*The below diagram shows Thing 1 is connected to channel 1 and channel 2 and Thing 2 is connected to Channel 2. This relationship can be established using the provided request*_
```

Also there is no provided request

Author: Starting part of document explain about the hierarchy, Assign and unassign of entities are explained in later part of document

```mermaid
graph
   subgraph Domain_1
      direction BT
      Gr1["Group 1"]
      Gr2["Group 2"]

      Th1["Thing 1"]
      Th2["Thing 2"]

      Ch1["Channel 1"]
      Ch2["Channel 2"]

      Th1 --->|connect| Ch1
      Th1 --->|connect| Ch2

      Th2 --->|connect| Ch2
   end
```

##### Channel Group Relation

> A Group serves as a parent entity that can contain both child Groups and Channels. Child Groups, in turn, can consist of further child Groups or Channels, forming a nested hierarchy. Notably, Channels, which are distinct entities, cannot have child Channels but can connect to multiple Things. The concept of parentage signifies the relationship between higher-level entities and their subordinate components. Ancestors in this system refer to entities higher up in the hierarchy, and while a child group can have multiple ancestors, a Channel can only belong to a single parent Group. This hierarchical arrangement provides a structured and organized framework for managing information within the Magistrala.

Reviewer: ```suggestion
A Group serves as a parent entity that can contain both groups and channels as children.
```
Same as other cases

Assigning a group as the parent of a channel can be achieved through the following request.

> ```bash

Reviewer: same

curl --location 'http://localhost/channels/<Channel 1 ID>/groups/assign' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer <DOMAIN_USER_ACCESS_TOKEN>' \
--data '{
    "group_ids" : [ "<Group 2 ID>" ]
}'
```

> _*The diagram below illustrates the parent relationship between Channel 1 and Channel 2 with Group 2. This relationship can be established using the provided request.*_

Reviewer: same

style Domain stroke-width:3px,margin-top:10px,margin-bottom:10px
```

#### Domain Entities Relations

> Domain hold all entities such as Groups, Channels and Things.

Reviewer 2: ...holds entities (not all, e.g. Users).

> The entities created in Domain doesn't creates have hierarchical structure within domain.

Reviewer 2: This is confusing, please fix it.

> Example: In Domain_1 a user creates following entities Group 1, Group_2, Thing 1, Thing 2, Channel 1, Channel 2. By default their will be relation betweens the entities, until user assigns relation between entities

Reviewer 2: `By default their will be relation betweens the entities, until user assigns relation between entities` What does it mean: Will there be an implicit relation or not?

```mermaid
graph
   subgraph Domain_1
      direction TB
      Gr1["Group 1"]
      Gr2["Group 2"]

      Th1["Thing 1"]
      Th2["Thing 2"]

      Ch1["Channel 1"]
      Ch2["Channel 2"]

   end
```

##### Channel Thing Connect/Disconnect

> `Thing` represents devices (or applications) connected to Magistrala that uses the platform for message exchange with other `things`.

Reviewer 2: ...a device (or an application)...

> `Channel` represents a communication channel. It serves as message topic that can be consumed by all of the things connected to it.

Reviewer 2: ... as a message conduit between things connected to it.

> To Channel communication topics things can publish/Subscribe the messages.

Reviewer 2: Things can publish or subscribe to the channel.

Thing and Channel can be connected to multiple channels using the following API.

```bash
curl --location 'http://localhost/connect' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'Authorization: Bearer <DOMAIN_USER_ACCESS_TOKEN>' \
--data '{
  "thing_id": "<Thing 1 ID>",
  "channel_id": "<Channel 1 ID>"
}'
```

_*The below diagram shows Thing 1 connect to channel 1 , channel 2 and Thing 2 connect to Thing 2. This relationship can be established using the provided request*_

```mermaid
graph
   subgraph Domain_1
      direction BT
      Gr1["Group 1"]
      Gr2["Group 2"]

      Th1["Thing 1"]
      Th2["Thing 2"]

      Ch1["Channel 1"]
      Ch2["Channel 2"]

      Th1 --->|connect| Ch1
      Th1 --->|connect| Ch2

      Th2 --->|connect| Ch2
   end
```

##### Channel Group Relation

> Groups are entities that can be linked as parents to channels.

Reviewer 2: Explain (what does parent mean, how does hierarchy work, can there be multiple ancestors...).

}
```

In JWT token, the domain field has **domain ID** and the user field has **user ID**.

If the domain field is empty, then with that JWT token following actions are permitted

- User profile update
- Domain creation & listing,
- Accept domain invitations

> Actions related to the creation, updation, and deletion of things, channels, and groups are not permitted, requests will fail in authorization. In magistrala operations related to things, channels, and groups take place in domain level. So for these kinds of operations, a JWT token with a domain field containing the operating domain ID is required.

Reviewer 2: ...creation, update, and deletion...

Reviewer 2: @arvindh123 Please address this one, as well.

> Actions related to the creation, updation, and deletion of things, channels, and groups are not permitted, requests will fail in authorization. In magistrala operations related to things, channels, and groups take place in domain level. So for these kinds of operations, a JWT token with a domain field containing the operating domain ID is required.

Reviewer 2: Also, use capital and change: _~In~ Magistrala operations related to things..._
